designs/2019-worker-threads/README.md

```diff
@@ -0,0 +1,338 @@
+     1. Workers [verify files]. It reports the lint result to master, and master may send the next file to the worker. The worker verifies the received file.
+     1. Workers exit with `0` after it reported the result of all files.
+1. Master collects the lint result from workers then write it to the lint result cache.
+1. Master returns the merged result from cache and workers.
+
+#### Worker's implementation
+
+Worker can be `require("worker_threads").Worker` or `require("child_process").ChildProcess`.
+
+If `v12.11.0` or newer version of Node.js runs ESLint, it's `require("worker_threads").Worker`. Otherwise it's `require("child_process").ChildProcess`.
+
+#### About `--cache` option
+
+We can use parallel linting along with `--cache` option because worker threads don't touch the cache. Master gives worker threads only files which are not cached (or cache was obsolete) and updates the cache with the collected result.
+
+#### About `--fix` option
+
+We can use parallel linting along with `--fix` option. ESLint modifies files after all linting completed.
+
+#### About `--debug` option
+
+We can use parallel linting along with `--debug` option, but ESLint cannot guarantee the order of the debug log. The order of the logs that are received from worker threads is messed.
+
+#### About Error Handling
+
+If exceptions happen in worker threads, the worker transfers the error information to the main thread if possible. The transferred information contains the properties of `message`, `stack`, `messageTemplate`, and `messageData`.
+
+Therefore, people can see the error message as-is even if it happened in worker threads.
+For example:
+
+```
+$ eslint lib
+Error: FAILED!!!
+Occurred while linting C:\Users\t-nagashima.AD\dev\eslint\lib\rules\no-mixed-operators.js:229
+    at BinaryExpression (C:\Users\t-nagashima.AD\dev\eslint\lib\rules\eqeqeq.js:152:27)
+    at C:\Users\t-nagashima.AD\dev\eslint\lib\linter\safe-emitter.js:45:58
+    at Array.forEach (<anonymous>)
+    at Object.emit (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\safe-emitter.js:45:38)
+    at NodeEventGenerator.applySelector (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\node-event-generator.js:253:26)
+    at NodeEventGenerator.applySelectors (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\node-event-generator.js:282:22)
+    at NodeEventGenerator.enterNode (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\node-event-generator.js:296:14)
+    at CodePathAnalyzer.enterNode (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\code-path-analysis\code-path-analyzer.js:646:23)
+    at C:\Users\t-nagashima.AD\dev\eslint\lib\linter\linter.js:936:32
+    at Array.forEach (<anonymous>)
+```
+
+As a side note, master terminates the other workers if a worker reported an error.
+
+#### About aborting
+
+In [RFC 40](https://github.com/eslint/rfcs/pull/40), `ESLint#executeOnFiles()` returns async iterable object.
```

Comment:
No, an iterable object is not an iterator.

- Async iterable protocol has the `[Symbol.asyncIterator]()` method that returns an async iterator object. We can use the iterable with `for-await-of` loops.
- Async iterator protocol has the `next()` method to iterate values, and has optional `return()` and `throw()` methods to operate the state of the iterator.  We **cannot** use the iterator with `for-await-of` loops.

What the `ESLint#executeOnFiles()` method returns is an iterable object because we want to use it with `for-await-of` loops.